### DISCUSSION TITLE
Using dioxus / leptos in combination with loco

### DISCUSSION THREAD

**Discussion by ckrenslehner** [loco-rs/loco#1068] [Q&A]
Hi! 

Disclaimer: I do not have any experience in using dioxus/leptos/loco. I am looking at all these options and just want to wrap my head around when to use what. 

As I understand from the docs, both leptos and dioxus offer “server functions” which make them kind of fullstack. However, loco seems to be a very good choice for a battery included backend framework. 
So a general question: Does it make any sense to even use something like leptos/dioxus or is it just an overkill?

Sorry if this question does not even make sense, and thanks for your feedback :-)

Greetings, Christian

**Comment by jondot** [comment]
Hi.  👋🏻 
It’s a good question. I went into trying out both myself in the past. 
my conclusion is that I can be much more productive with Loco as the backend because it has a mindset and I can wrap in my head easily. You have a server, models, controllers, and a lot of developer experience tooling you need in real life. 

So now what remains is the frontend part of these frameworks. For example you can take leptos, and build a client app with it and connect to a loco backend. This is easy since the leptos part can be a standalone crate, you build it and put the assets in loco’s assets folder (it would be some wasm and some js). 

And then the question remains are you comfortable operating wasm based client apps.  You can build them with Rust for sure and that will be fun. But can you operate them? Will you have the same debugging tools as say React has, same monitoring, same testing tools, etc? And will it matter for your case?

In my mind this is what it comes to

**Reply by ckrenslehner** [reply]
Hi! Thanks for your detailed answer. I am not very familiar with web programming in general, but want to get into it. As I am very hyped about rust for various reasons, I am interested in using rust all the way. As I understand, leptos and dioxus both use a component model similar to the concepts of React. So I guess I can at least take some general knowledge with me :-)
A concrete project I am thinking about is a "simple" website where users can upload binary files. The server should act as a tcp/udp packet server for over the air firmware downloads for embedded devices. So I need support for authentication, tcp/udp and protobuf.
What I really like about loco is the mailer module, which seems a very big advantage in regard to user authentication. :-)

Long story short, I currently do not have any testing/monitoring/debugging tools, but of course I appreciate it if you can give me some insights!

**Comment by brandonpollack23** [comment]
I've made a prototype of this: https://github.com/brandonpollack23/loco_dioxus_test

and there is some work needed on teh dioxus end but maybe on the loco end to make some stuff easier like if there was a more robust way to access the axum router before it is "done" we could just use the regular dixous server initializer function on it instead of recreating a bunch of stuff.

The majority of the work is in a controller that is relevant to loco.rs I'd love to know what you think @jondot

**Reply by brandonpollack23** [reply]
FYI I've also started a chat abbout this on the dioxus discord.  I would LOVE it if you guys could suggest stuff to make this a part of loco and dioxus :)

https://discordapp.com/channels/899851952891002890/899851952891002893/1451411960275927123
